src/singleprover.cpp

```diff
@@ -0,0 +1,127 @@
+#include <singleprover.hpp>
+#include "fr.hpp"
+
+#include "logger.hpp"
+#include "wtns_utils.hpp"
+
+SingleProver::SingleProver(std::string zkeyFileName) {
+    LOG_INFO("SingleProver::SingleProver begin");
+    auto t0 = std::chrono::steady_clock::now();
+
+    mpz_init(altBbn128r);
+    mpz_set_str(altBbn128r, "21888242871839275222246405745257275088548364400416034343698204186575808495617", 10);
+
+    zKey = BinFileUtils::openExisting(zkeyFileName, "zkey", 1);
+    zkHeader = ZKeyUtils::loadHeader(zKey.get());
+
+    std::string proofStr;
+    if (mpz_cmp(zkHeader->rPrime, altBbn128r) != 0) {
+        throw std::invalid_argument( "zkey curve not supported" );
+    }
+    
+    prover = Groth16::makeProver<AltBn128::Engine>(
+        zkHeader->nVars,
+        zkHeader->nPublic,
+        zkHeader->domainSize,
+        zkHeader->nCoefs,
+        zkHeader->vk_alpha1,
+        zkHeader->vk_beta1,
+        zkHeader->vk_beta2,
+        zkHeader->vk_delta1,
+        zkHeader->vk_delta2,
+        zKey->getSectionData(4),    // Coefs
+        zKey->getSectionData(5),    // pointsA
+        zKey->getSectionData(6),    // pointsB1
+        zKey->getSectionData(7),    // pointsB2
+        zKey->getSectionData(8),    // pointsC
+        zKey->getSectionData(9)     // pointsH1
+    );
+
+    auto t1 = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count();
+    std::string output("SingleProver::SingleProver initialized from zkey in " + std::to_string(duration) + "ms");
+    LOG_INFO(output);
+}
+
+SingleProver::~SingleProver()
+{
+    mpz_clear(altBbn128r);
+}
+
+#include <unistd.h>
+
+// Function to generate a unique temporary file name
+std::string generateTempFileName() {
+    char templateName[] = "/tmp/zklogin_XXXXXX";
+    int fd;
+    if ((fd = mkstemp(templateName)) == -1) {
```

Comment:
let's use c++17 stdlib's API instead

Reply:
The other alternative I've briefly looked at was [tempFile](https://en.cppreference.com/w/cpp/io/c/tmpfile). But it was not clear how to access the filename with that. 

Did you have any specific alternative in mind? And what is the problem with `mkstemp`?

Reply:
mainly because it's a posix/C interface and not C++ so I'd guess the latter is better, and, we don't actually use the only advantage of mkstemp (atomic create and open). that said, if there is no simple alternative, let's stay with mkstemp

Reply:
Yeah, I'll stay with it for now. I did spend some time looking at it. Will add a TODO to look again later.